chalk-solve/src/coherence.rs

```diff
@@ -106,18 +106,24 @@ where
 
     // Build the forest of specialization relationships.
     fn build_specialization_forest(&self) -> Result<Graph<ImplId<I>, ()>, CoherenceError<I>> {
-        // The forest is returned as a graph but built as a GraphMap; this is
-        // so that we never add multiple nodes with the same ItemId.
-        let mut forest = DiGraphMap::new();
+        let mut forest = DiGraph::new();
+
+        let node_impls: Vec<ImplId<_>> = forest.raw_nodes().iter().map(|x| x.weight).collect();
 
         // Find all specializations (implemented in coherence/solve)
         // Record them in the forest by adding an edge from the less special
         // to the more special.
         self.visit_specializations_of_trait(|less_special, more_special| {
-            forest.add_edge(less_special, more_special, ());
+            // Check so that we never add multiple nodes with the same ImplId.
+            if !node_impls.contains(&less_special) && !node_impls.contains(&more_special) {
```

Comment:
This seems wrong. Theoretically, we can have these two nodes, but not connected to each other.

Isn't [`update_edge`](https://docs.rs/petgraph/0.6.0/petgraph/graph/struct.Graph.html#method.update_edge) the right function to call? "Note: Graph allows adding parallel (“duplicate”) edges. If you want to avoid this, use .update_edge(a, b, weight) instead."

Reply:
Testing this is https://github.com/rust-lang/chalk/pull/740/commits/c89e4b79465bc9fe7e17b6f9ed42e08c79137e08.